inginious/agent/docker_agent/__init__.py

```diff
@@ -545,14 +564,30 @@ async def handle_running_container(self, info: DockerRunningJob, future_results)
                                 hard_time_limit = min(msg["hard_time_limit"] or info.hard_time_limit, info.hard_time_limit)
                                 share_network = msg["share_network"]
                                 socket_id = msg["socket_id"]
+                                ssh = msg["ssh"]
                                 assert "/" not in socket_id  # ensure task creator do not try to break the agent :-(
                                 self._create_safe_task(self.create_student_container(info, socket_id, environment, memory_limit,
                                                                                      time_limit, hard_time_limit, share_network,
-                                                                                     write_stream))
-                            elif msg["type"] == "ssh_key":
-                                # send the data to the backend (and client)
+                                                                                     write_stream, ssh))
+                            elif msg["type"] == "ssh_debug":
+                                # send the data to the frontend (and client) to reach grading_container
                                 self._logger.info("%s %s", info.container_id, str(msg))
                                 await self.send_ssh_job_info(info.job_id, self._address_host, info.ports[22], msg["ssh_user"], msg["ssh_key"])
+
+                            elif msg["type"] == "ssh_student":
+                                # send the data to the frontend (and client) to reach student_container
+                                if info.enable_network:
+                                    info_student = None
+                                    if len(self._student_containers_running) > 0 and msg["container_id"] in info.student_containers:
+                                        info_student = self._student_containers_running[msg["container_id"]]
+                                    else:
+                                        self._logger.exception("Exception: no linked student_container running.")
+                                    self._logger.info("%s %s", info_student.container_id, str(msg))
+                                    await self.send_ssh_job_info(info.job_id, self._address_host, info_student.ports[22], msg["ssh_user"], msg["ssh_key"])
+                                else:
+                                    await self.send_job_result(info.job_id, "ssh for student requires internet access in the task configuration !")
+                                    self._logger.exception("Exception: ssh for student requires internet access in the task configuration !.")
+                                    await self.kill_job(BackendKillJob(job_id=info.job_id))  # Just to be 100% sure the job is killed
```

Comment:
It seems that killing the job here may result in exceptions in ``handle_job_closing``:

    2021-06-30 08:48:54,638 - inginious.agent.docker - ERROR - Exception: ssh for student requires internet access in the task configuration !.
    NoneType: None
    2021-06-30 08:48:54,638 - inginious.backend - INFO - Job ad3d0b58-6368-4a64-97c1-25c14e388a35 finished on agent b'\x00\x19I]\x00'
    2021-06-30 08:48:54,713 - inginious.agent.docker - WARNING - Container 8dce2b6b4432639b526bde7e5acf2c2a378d65b1e9d60e132d64df9dff2d4969 has not given any result
    2021-06-30 08:48:54,723 - inginious.agent.docker - ERROR - Exception in handle_job_closing
    Traceback (most recent call last):
      File "/home/anthony/Documents/inginious-dev/INGInious/inginious/agent/docker_agent/__init__.py", line 781, in handle_job_closing
        await self.send_job_result(info.job_id, result, error_msg, grade, problems, tests, custom, state, archive, stdout, stderr)
      File "/home/anthony/Documents/inginious-dev/INGInious/inginious/agent/__init__.py", line 215, in send_job_result
        raise JobNotRunningException()
    inginious.agent.JobNotRunningException
    2021-06-30 08:48:55,328 - inginious.webapp.requests - INFO - 127.0.0.1:48590 - "- POST /course/test/test_ssh" - 200 OK
    127.0.0.1 - - [30/Jun/2021 08:48:55] "POST /course/test/test_ssh HTTP/1.1" 200 -

Moreover, I don't know where this ``NoneType: None`` comes from in my logs. The error message is not displayed in the webapp. I'm only left with *Internal error* :-(

Reply:
Indeed, the error should be correctly handled and the feedback should be more explicit.
The strange NoneType: None origin is surprisingly hard to find :o 
Working on this right now :)

Reply:
About the "Nonetype: None" being logged, it's clearly the self._logger.exception("...") that makes this happen...
But no idea why :/